[Feature/recruitment serializers] recruitment_bookmark api작성

Yong-back opened the pull request:
## ✅ PR 요약
- 작업 요약: recruitment_bookmark serializer , views, test_code 작성

## 📄 상세 내용
- [x] recruitment_bookmark
- [x] serializer 작성 
- [x] views 작성
- [x] test_code 작성

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Yong-back committed:
✨ feat: recruitment_bookmarks serializer 작성

recruitment_bookmarks 의 serializer 작성하였습니다.

Yong-back committed:
✨ feat: recruitment_search_logs serializer 작성

recruitment_bookmarks serializer 작성 후 recruitment_search_logs serializer 작성 완료

Yong-back committed:
✨ feat: recruitment_bookmark view 작

recruitment_bookmark view.py 작성
성

Yong-back committed:
✨ feat: recruitment_bookmark 해당하는 view , test_code 작성

recruitment_bookmark 해당 views, test_code , serializer 작성 완료

SWWS97 commented on apps/recruitment/serializers/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,113 @@
+        )
+
+    def get_thumbnail_img_url(self, obj: "RecruitmentBookmarks") -> Optional[str]:
+        recruitment = getattr(obj, "recruitment_id", None)
```
recruitment_id가 아니라 recruitment 수정해주세요

SWWS97 commented on apps/recruitment/serializers/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,113 @@
+
+    #  예상 모집 인원
+    expected_member_count = serializers.IntegerField(
+        source="recruitment.expected_member_count",
```
이부분도 명세서랑 다른데 혹시 수정을 안하신걸까요? 명세서엔 expected_headcount로 되어 있어서요

SWWS97 commented on apps/recruitment/serializers/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,113 @@
+
+    #  마감 기한
+    close_at = serializers.DateTimeField(
+        source="recruitment.deadline",
```
여기도 마찬가지로 deadline이 아니라 close_at이고

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,137 @@
+        user = request.user
+
+        try:
+            bookmark, created = RecruitmentBookmarks.objects.get_or_create(
```
get_or_create에서 이미 있으면 가져오고 없으면 생성인데 왜 IntegrityError가 있을까요?

Yong-back replied:
동시성 때문에 일단 try except 로 했는데 확률적으로 드물거 같으니까 일단 그냥 get_or_create 만 사용하는 쪽으로 수정하겠습니다

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,137 @@
+        description="Cursor Pagination 기반 북마크 목록 조회, 타이틀 검색 가능",
+        responses={200: RecruitmentBookmarkCardSerializer},
+    )
+    def get(self, reqeust: Request, *args: Any, **kwargs: Any) -> Response:
```
여기 reqeust 오타 났어요

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,137 @@
+    def get(self, reqeust: Request, *args: Any, **kwargs: Any) -> Response:
+        queryset = self.get_queryset()
+        paginator = self.pagination_class()
+        page = paginator.paginate_queryset(queryset, self.request)
```
여기도 self.request로 가져오는게 없는데 request로 수정하셔야 할거 같습니다

SWWS97 commented on apps/recruitment/serializers/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,113 @@
+
+    #  북마크 수
+    bookmark_count = serializers.IntegerField(
+        source="recruitment.bookmark_count",
```
bookmark에 count 필드가 없는데 어떻게 가져오시는거죠...

Yong-back replied:
annotate 로 수정하였고 , 
* 현재 로그인 유저 기준으로 북마크 쿼리셋이 생성됩니다.

Yong-back committed:
bookmark serializer
-> bookmark_count source 제거 후 annotate 로 view 에서 들고올수있는 형식으로 view 수정

Yong-back commented:
코멘트 확인 후 수정하였습니다

SWWS97 commented on apps/recruitment/serializers/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,112 @@
+    )
+
+    #  조회수
+    view_count = serializers.IntegerField(
```
여기 recuitment 보시면 필드에 views_count라고 되어있어요

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,141 @@
+            return RecruitmentBookmarks.objects.none()
+
+        queryset = RecruitmentBookmarks.objects.select_related("recruitment_id").annotate(
+            is_bookmarked_by_user=Exists(user_bookmarks.filter(recruitment_id=OuterRef("recruitment_id"))),
```
여기 아래에  bookmark_count=Count("recruitment_id__recruitment_bookmarks") 이것도 추가하셔야 해요

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,141 @@
+        return paginator.get_paginated_response(serializer.data)
+
+    def get_queryset(self) -> QuerySet[RecruitmentBookmarks]:
+        user = self.request.user
```
여기 self.request.user -> request.user입니다

Yong-back committed:
💡 chore    : views ,serializer 코드 수

serializer 내에 view_count -> views_count 로 수정
views 코드 bookmark_count 에 해당하는 내용 수정 및
self.request.user -> request.user 로 수정
정

Yong-back commented:
코멘트 반영하여 수정하였습니다.

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,142 @@
+
+        user_bookmarks = RecruitmentBookmarks.objects.filter(user_id=user.id)
+
+        if user.is_anonymous:
```
여기서 이미 user.is_anonymous 검증 하고 있으니까 위에는 빼셔도 될거 같아요 중복이라

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,142 @@
+        ],
+    )
+    def post(self, request: Request, *args: Any, **kwargs: Any) -> Response:
+        # if request.user.is_anonymous:
```
여기 주석지우고 올려주세용

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,142 @@
+        # if request.user.is_anonymous:
+        #     return Response({"detail": "자격 인증 데이터가 필요합니다."}, status=status.HTTP_401_UNAUTHORIZED)
+
+        serializer_class = self.get_serializer_class()
```
self.get_serializer_class() request 전달 해주셔야해요 self.get_serializer_class(request)

SWWS97 commented on apps/recruitment/views/recruitment_bookmarks.py:
```diff
@@ -0,0 +1,142 @@
+            return RecruitmentBookmarks.objects.none()
+
+        queryset = RecruitmentBookmarks.objects.select_related("recruitment_id").annotate(
+            is_bookmarked_by_user=Exists(user_bookmarks.filter(recruitment_id=OuterRef("recruitment_id"))),
```
여기서 이미 북마크 자체가 유저가 북마크 했을때니까 유저가 북마크 했는지를 검증할 필요가 없으니까 빼야할거같아요
어차피 항상 True니까 그리고 지금 전체 북마크를 들고 오는데 유저가 북마크 한거를 들고 와야하니까 유저id로 필터링해서
가져오고 annotate 사용하시면 될거같아요

Yong-back committed:
💡 chore    : recruitment_bookmark view 수정

recruitment_bookmark view 수정
